File: graphs/dfs.cpp
#include<bits/stdc++.h>
using namespace std;

// 'u' is a node 'from', 'v' is a node 'to': from u to v
void add_edge_list(vector<vector<int>> &g, int u, int v) {
    g[u].push_back(v);      
    g[v].push_back(u);      // Delete this line if graph is directed
}

void add_edge_matrix(vector<vector<int>> &g, int u, int v) {
    g[u][v] = 1;
    g[v][u] = 1;            // Delete this line for directed graph
}

// Two functions to perform the DFS

// LIST
void dfs_list_visit(vector<vector<int>> &g, int start, bool* visited){
    stack <int> s;
    
    s.push(start);
    visited[start] = true;

    while (!s.empty()) {
        int u = s.top(); s.pop();
        cout << u << " ";
        for (auto v: g[u]) {
            if (!visited[v]) {
                s.push(v);
                visited[v] = true;
            }
        }
    }
    cout << endl;
}

void dfs_list(vector<vector<int>> &g) {
    cout << "DFS LIST:\n";

    int n = g.size();
    bool* visited = new bool[n]();

    for (int i = 0; i < n; i++) {
        if (!visited[i])
            dfs_list_visit(g, i, visited);
    }
}

// MATRIX
void dfs_matrix_visit(vector<vector<int>> &g, int start, bool* visited) {
    stack<int> s;

    s.push(start);
    visited[start] = true;

    while (!s.empty()) {
        int u = s.top(); s.pop();
        cout << u << " ";

        for (int v = 0; v < g.size(); v++) {
            // if (start == v)
            //     continue;
            if (g[u][v] == 1 && !visited[v]) {
                s.push(v);
                visited[v] = true;
            }
        }
    }

    cout << endl;
}

void dfs_matrix(vector<vector<int>> &g) {
    cout << "DFS MATRIX:\n";

    int n = g.size();
    bool* visited = new bool[n]();

    // This for loop is only beneficial when your graph is disconnected
    for (int i = 0; i < n; i++) {
        if (!visited[i])
            dfs_matrix_visit(g, i, visited);
    }

    // If connected, you can do it in single call:
    // dfs_matrix_visit(g, 2, visited);
}

int main() {
    int n, e; cin >> n >> e;

    // Adjacency list
    vector<vector<int>> adj_list(n, vector<int>());

    // Adjacency matrix
    vector<vector<int>> adj_matrix(n, vector<int>(n, 0));   // nxn matrix all zeros

    // Our nodes are from 0 to n-1
    while (e--) {
        int u, v; cin >> u >> v;
        add_edge_list(adj_list, u, v);
        add_edge_matrix(adj_matrix, u, v);
    }

    dfs_list(adj_list);
    dfs_matrix(adj_matrix);

    return 0;
}

File: graphs/dfs.py
def add_edge_list(g, u, v):
    g[u].append(v)
    g[v].append(u)  # Delete this line if the graph is directed

def add_edge_matrix(g, u, v):
    g[u][v] = 1
    g[v][u] = 1  # Delete this line for directed graph

def dfs_list_visit(g, start, visited):
    stack = []
    stack.append(start)
    visited[start] = True

    while stack:
        u = stack.pop()
        print(u, end=" ")
        for v in g[u]:
            if not visited[v]:
                stack.append(v)
                visited[v] = True
    print()

def dfs_list(g):
    print("DFS LIST:")
    n = len(g)
    visited = [False] * n

    for i in range(n):
        if not visited[i]:
            dfs_list_visit(g, i, visited)

def dfs_matrix_visit(g, start, visited):
    stack = []
    stack.append(start)
    visited[start] = True

    while stack:
        u = stack.pop()
        print(u, end=" ")
        for v in range(len(g)):
            if g[u][v] == 1 and not visited[v]:
                stack.append(v)
                visited[v] = True
    print()

def dfs_matrix(g):
    print("DFS MATRIX:")
    n = len(g)
    visited = [False] * n

    for i in range(n):
        if not visited[i]:
            dfs_matrix_visit(g, i, visited)

def dfs_visit(g, u, visited):
    visited[u] = True
    print(u, end=" ")

    for v in g[u]:
        if not visited[v]:
            dfs_visit(g, v, visited)

def dfs(g):
    print("RECURSIVE DFS LIST:")
    n = len(g)
    visited = [False] * n

    for i in range(n):
        if not visited[i]:
            dfs_visit(g, i, visited)
    print()

def main():
    n, e = map(int, input().split())

    # Adjacency list
    adj_list = [[] for _ in range(n)]

    # Adjacency matrix
    adj_matrix = [[0] * n for _ in range(n)]

    # Our nodes are from 0 to n-1
    for _ in range(e):
        u, v = map(int, input().split())
        add_edge_list(adj_list, u, v)
        add_edge_matrix(adj_matrix, u, v)

    dfs_list(adj_list)
    dfs_matrix(adj_matrix)
    dfs(adj_list)

if __name__ == "__main__":
    main()
